onionshare/onionshare
# Fetch the built-in bridges from Tor's Censorship Circumvention API, rather than hardcode them

As part of the anti-censorship project.

## Change

Instead of hardcoding copies of the obfs4, snowflake and meek_lite bridges (which can go out of date easily), we can now use the new Censorship Circumvention API at Tor Project, which has a route that provides the latest built-in bridges.

(Tor Project will be using this themselves so as not to have to ship hardcoded built-in bridges with Tor Browser, so this is the same sort of change here).

As usual, the request to the censorship circumvention API is done via Meek domain-fronting.

## Things to consider

One quirk of this, is that if the user has enabled built-in bridges, then each subsequent time they open OnionShare, it will start Meek, connect to the Censorship Circumvention API, fetch the latest built-in bridges, write them to the torrc and then connect. 

This is a double-edged sword:

Pro: it means each time the user opens onionshare, they get the latest built-in bridges, so are less likely to be using an obsolete bridge that went offline since last time.

Con: it means their machine always performs a non-Tor-connected domain-fronted request to the Tor API, which might be fingerprint-able.

What do we think about that? Would you rather have it *periodically* write the bridges to a torrc template, e.g random N times out of 10 connect back to Tor API to get the latest bridges? This would be more complex..


## Steps to test

1. Start OnionShare with `./scripts/dev.sh -v`
2. Choose 'Use a Bridge' -> 'Select a built-in bridge' -> obfs4, or snowflake, or meek-azure
3. Click Save
4. Whilst Tor is reconnecting, you could cat the `~/.config/onionshare/tor_data/torrc` and you should see the Tor bridges get written to the torrc. In addition you should see some debug data in the console that shows it fetching the json blob of built-in bridges from the Tor API.

I noted that meek-azure is timing out for me at 2%, but I think this is not our problem, just a problem with that bridge type. It was definitely executing the obfs4proxy binary to provide the meek_client bridge type. But obfs4 and snowflake worked great for me.
